### ISSUE TITLE
FastCGM errors when running locally

### ISSUE THREAD

**Issue by eliasen** [superpermutators/superperm#42]
I've compiled FastCGM maybe on Fedora 30 Linux (see my other bugs for compiler flag fixes) and when I run locally on the machine with the graphics cards, I get the error:
```

Random seed is: 1562845468
Thu Jul 11 02:15:57 2019 Program instance number: 1130860108
X server found. dri2 connection failed! 
DRM_IOCTL_I915_GEM_APERTURE failed: Invalid argument
Assuming 131072kB available aperture size.
May lead to reduced performance or incorrect rendering.
get chip id failed: -1 [22]
param: 4, val: 0
Found 3 OpenCL platforms on this system:
Platform 1:
Profile: FULL_PROFILE
Version: OpenCL 2.0 beignet 1.3
Name: Intel Gen OCL Driver
Vendor: Intel
X server found. dri2 connection failed! 
DRM_IOCTL_I915_GEM_APERTURE failed: Invalid argument
Assuming 131072kB available aperture size.
May lead to reduced performance or incorrect rendering.
get chip id failed: -1 [22]
param: 4, val: 0
cl_get_gt_device(): error, unknown device: 0
OpenCL returned the error: CL_DEVICE_NOT_FOUND

```
Using the "inxi" program, my graphics configuration is:

```
inxi -GS
System:
  Host: snoopy Kernel: 5.1.15-300.fc30.x86_64 x86_64 bits: 64 
  Desktop: MATE 1.22.1 Distro: Fedora release 30 (Thirty) 
Graphics:
  Device-1: NVIDIA GM206 [GeForce GTX 960] driver: nvidia v: 430.26 
  Display: x11 server: Fedora Project X.org 1.20.5 driver: nvidia 
  unloaded: fbdev,modesetting,nouveau,vesa resolution: 1920x1080~60Hz 
  OpenGL: renderer: GeForce GTX 960/PCIe/SSE2 v: 4.6.0 NVIDIA 430.26
```

**Comment by nagegerg** [comment]
Thanks. There's an awful lot of messages here that aren't coming from the FastDCM program itself, and if you're getting error messages (from X11? From some driver?) when all FastDCM has done is query what OpenCL platforms are available, that makes me suspect that the OpenCL support on this system is broken.

**Comment by eliasen** [comment]
Yeah, I'm not sure how to proceed.  I deleted the Mesa drivers which went from "Found 4 OpenCL 
platforms on this system" to 3, and my main driver is NVIDIA.  I did try:

`./FastDCM gpuName "NVIDIA"`

which gives the same errors.  I'll try to figure out how to proceed.

**Comment by nagegerg** [comment]
Let me get back to you with a new option for selecting the OpenCL platform. At the moment, gpuName only filters on the device names (devices are lower down in the hierarchy than platforms), whereas I think we're going to have to screen out the whole Intel OpenCL **platform** to find out if that's the problem.

**Comment by eliasen** [comment]
I've done some searching and the line

`Version: OpenCL 2.0 beignet 1.3`

leads me to believe that beignet is the "Open source implementation of the OpenCL for Intel GPUs" which I don't even know if I have.  Do I have an Intel GPU?   Maybe the base chipset on my motherboard?  I'm tryiing 

`dnf remove beignet`

on Fedora 30 to see if this gets me closer.

**Comment by eliasen** [comment]
After I do:

```
./FastDCM gpuName "NVIDIA"
Random seed is: 1562849641
Thu Jul 11 03:05:26 2019 Program instance number: 70092256
Found 2 OpenCL platforms on this system:
Platform 1:
Profile: FULL_PROFILE
Version: OpenCL 1.2 pocl 1.3-pre RelWithDebInfo, LLVM 8.0.0, SLEEF, DISTRO, POCL_DEBUG
Name: Portable Computing Language
Vendor: The pocl project
OpenCL returned the error: CL_DEVICE_NOT_FOUND
eliasen@snoopy FastDCM $ ./FastDCM
Random seed is: 1562849676
Thu Jul 11 03:05:30 2019 Program instance number: 1239477934
Found 2 OpenCL platforms on this system:
Platform 1:
Profile: FULL_PROFILE
Version: OpenCL 1.2 pocl 1.3-pre RelWithDebInfo, LLVM 8.0.0, SLEEF, DISTRO, POCL_DEBUG
Name: Portable Computing Language
Vendor: The pocl project
OpenCL returned the error: CL_DEVICE_NOT_FOUND
```

So now I'm battling against the "pocl" driver which is apparently this:
http://portablecl.org/

Doing a " dnf remove pocl" gives me this:
```
Removing:
 pocl          x86_64   1.2-4.20190221gita0b083a1b47a738.fc30   @fedora    57 M
Removing unused dependencies:
 clang-libs    x86_64   8.0.0-1.fc30                            @fedora   111 M
 compiler-rt   x86_64   8.0.0-1.fc30                            @fedora    32 M
 hwloc-libs    x86_64   1.11.12-2.fc30                          @fedora   2.2 M
```

And after removing those (I hope I don't need them} then I seem to be able to run FastCGM

**Comment by nagegerg** [comment]
Wow, that's some jungle you've scythed your way through!

**Comment by eliasen** [comment]
Thanks!  I'm a novice at OpenCL so I'm hoping that we can find a way to find the *real* OpenCL driver on the system to make this easier for more people.  I have no idea how to do that, though.  

As noted before, the command

`inxi -GS`

seemed to find the right driver.   It may be here:
https://github.com/smxi/inxi

GPU programming is deep dark wizardry, though, and I'm really impressed that you've gotten it working.  Especially across platforms.

**Comment by nagegerg** [comment]
The new command-line option

gpuPlatform XXXX

should filter out any OpenCL platform offering itself to the program unless either the platform's name or the platform's vendor name start with XXXX. I don't know how common it is for "extra" implementations of OpenCL to be lying around on Linux installations, but this should at least make it less work to single out the one that's wanted.

**Comment by nagegerg** [comment]
I've now changed FastDCM so it will simply skip over any platforms that return an error when asked what GPUs they control, rather than crashing out of the program completely. I think that should get around the problem of the user trying to figure out what platform to select.

Thanks for all your help!
